### PR TITLE
global sort: enable concurrent read test

### LINKED ISSUE #49537: import into failed with tidb log error“read xxx bytes from external storage, exceed max limit 1073741824” when inject network latency form s3 (global sort dir)
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
1、open global sort and tidb_enable_dist_task
2、import into
3、inject network latency form s3 (global sort dir)

### 2. What did you expect to see? (Required)
import into success

### 3. What did you see instead (Required)

the status of import job is not finished or running (now: 2023-12-16 04:32:37, jobId: 150003, step: importing, status: failed)
operatorLogs:
[2023-12-16 04:06:07] ###### start import into
[2023-12-16 04:06:07] ###### wait for import job to finish
[2023-12-16 04:32:37] ###### wait for import job to finish failed
select id, step, status from mysql.tidb_import_jobs where start_time >= '2023-12-16 04:06:07'
jobId: 150003, step: importing, status: failed

tidb logs:
import into failed with log error“read xxx bytes from external storage, exceed max limit 1073741824”

[2023/12/16 04:32:06.313 +08:00] [ERROR] [reader.go:48] ["read all data failed"] [takeTime=1m2.635809773s] [error="read 3772936486386853641 bytes from external storage, exceed max limit 1073741824"]
[2023/12/16 04:32:06.313 +08:00] [ERROR] [local.go:1710] ["do import meets error"] [error="read 3772936486386853641 bytes from external storage, exceed max limit 1073741824"]

### 4. What is your TiDB version? (Required)
git hash：
074af1082d5a15019cb2720f54e942374ea83d1a

### PR DESCRIPTION
**Pull request by ywqzzy**
### What problem does this PR solve?

Issue Number: ref #48952 close #49537

Problem Summary:

### What changed and how does it work?
1. enable switch concurrent mode in byte reader test
2. fix a bug when switch mode.
### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: br/pkg/lightning/backend/external/byte_reader_test.go [hunk 1]
```diff
@@ -206,57 +208,63 @@ func TestEmptyContent(t *testing.T) {
 }
 
 func TestSwitchMode(t *testing.T) {
-	st, clean := NewS3WithBucketAndPrefix(t, "test", "testprefix")
-	defer clean()
-
+	seed := time.Now().Unix()
+	// seed := 1702880814
```
Comment (tangenta) [inline br/pkg/lightning/backend/external/byte_reader_test.go:212]:
```suggestion
```

File: br/pkg/lightning/backend/external/byte_reader.go [hunk 2]
```diff
@@ -313,15 +317,16 @@ func (r *byteReader) reload() error {
 func (r *byteReader) closeConcurrentReader() (reloadCnt, offsetInOldBuffer int) {
 	r.logger.Info("drop data in closeConcurrentReader",
 		zap.Int("reloadCnt", r.concurrentReader.reloadCnt),
-		zap.Int("dropBytes", len(r.curBuf)-r.curBufOffset),
+		zap.Int("dropBytes", len(r.curBuf[r.curBufIdx])-r.curBufOffset),
```
Comment (lance6716) [inline br/pkg/lightning/backend/external/byte_reader.go:320]: seems dropBytes need to add the totally unread curBuf `r.concurrentReader.bufSizePerConc * (len(r.curBuf) - 1 - r.curBufIdx)`

**Review by lance6716 — COMMENTED** [review]
rest lgtm

File: br/pkg/lightning/backend/external/byte_reader_test.go [hunk 3]
```diff
@@ -206,57 +208,62 @@ func TestEmptyContent(t *testing.T) {
...
-		if n == 0 {
-			break
-		}
-		y, err := br.readNBytes(n)
+		_, _, err := kvReader.nextKV()
 		if err == io.EOF {
 			break
 		}
 		require.NoError(t, err)
-		totalCnt += len(y)
```
Comment (tangenta) [inline br/pkg/lightning/backend/external/byte_reader_test.go:256]: Why not record `totalCnt`?

Comment (ywqzzy) [reply br/pkg/lightning/backend/external/byte_reader_test.go:256]: Not neccessary, I will compare all kvs.

**Review by lance6716 — APPROVED** [review]

**Review by D3Hunter — APPROVED** [review]

**Pull request merged**